android/library/src/test/java/io/logto/android/auth/browser/BrowserSignInFlowTest.kt

```diff
@@ -13,138 +13,152 @@ import io.logto.client.exception.LogtoException
+        } returns "signInUrl"
 
+        val activity: Activity = spyk(Robolectric.buildActivity(Activity::class.java).get())
         browserSignInFlow.start(activity)
 
-        verify(activity).startActivity(anyOrNull())
+        verify { activity.startActivity(any()) }
     }
 
     @Test
     @Suppress("UNCHECKED_CAST")
     fun startShouldInvokeCompleteWithLogtoExceptionOnDiscoverFailed() {
-        val mockLogtoException: LogtoException = mock()
-        doAnswer {
-            val block = it.arguments[0] as HandleOidcConfigurationCallback
+        val mockLogtoException: LogtoException = mockk()
+        coEvery {
+            logtoAndroidClientMock.getOidcConfigurationAsync(any())
+        } coAnswers {
+            val block = args[0] as HandleOidcConfigurationCallback
             block(mockLogtoException, null)
-            null
-        }.`when`(logtoAndroidClientMock).getOidcConfigurationAsync(anyOrNull())
-        val activity: Activity = spy(Robolectric.buildActivity(Activity::class.java).get())
+            Job(null)
+        }
 
+        val activity: Activity = spyk(Robolectric.buildActivity(Activity::class.java).get())
         browserSignInFlow.start(activity)
 
-        verify(onCompleteMock).invoke(logtoExceptionCaptor.capture(), tokenSetCaptor.capture())
-        assertThat(logtoExceptionCaptor.firstValue).isEqualTo(mockLogtoException)
-        assertThat(tokenSetCaptor.firstValue).isNull()
+        verify { onCompleteMock(capture(logtoExceptionSlot), isNull()) }
+        assertThat(logtoExceptionSlot.captured).isEqualTo(mockLogtoException)
     }
 
     @Test
     fun handleRedirectUriShouldCompleteWithLogtoExceptionWithEmptyUri() {
         val emptyUri = Uri.parse("")
-
         browserSignInFlow.handleRedirectUri(emptyUri)
 
-        verify(onCompleteMock).invoke(logtoExceptionCaptor.capture(), tokenSetCaptor.capture())
-        assertThat(logtoExceptionCaptor.firstValue)
+        verify { onCompleteMock.invoke(capture(logtoExceptionSlot), isNull()) }
+        assertThat(logtoExceptionSlot.captured)
             .hasMessageThat()
             .isEqualTo("$SIGN_IN_FAILED: $EMPTY_REDIRECT_URI")
-        assertThat(tokenSetCaptor.firstValue).isNull()
     }
 
     @Test
     fun handleRedirectUriShouldCompleteWithLogtoExceptionWithUriContainsErrorDescription() {
-        val invalidUri: Uri = mock()
-        `when`(invalidUri.getQueryParameter(eq(QueryKey.ERROR_DESCRIPTION)))
-            .thenReturn("mocked sign in error description")
+        val invalidUri: Uri = mockk()
+        every {
+            invalidUri.getQueryParameter(eq(QueryKey.ERROR_DESCRIPTION))
+        } returns "mocked sign in error description"
 
         browserSignInFlow.handleRedirectUri(invalidUri)
 
-        verify(onCompleteMock).invoke(logtoExceptionCaptor.capture(), tokenSetCaptor.capture())
-        assertThat(logtoExceptionCaptor.firstValue)
+        verify { onCompleteMock.invoke(capture(logtoExceptionSlot), isNull()) }
```

Comment:
shouldn't we verify toekSetCaptor.capture() is null  here?

Reply:
`isNull` is a matcher -- it asserts that `tokenSet` must be `null`.

Reply:
we should not assume the value we trying to capture is not null.
use `captureNullable`
```kotlin
val tokenSetSlot = mutableListOf<TokenSet?>()
val exceptionSlot = mutableListOf<LogtoException?>()

verify { onCompleteMock.invoke(captureNullable(exceptionSlot), captureNullable(tokenSetSlot)) }
assertThat(exceptionSlot.last())
    .hasMessageThat()
    .isEqualTo("$SIGN_IN_FAILED: $EMPTY_REDIRECT_URI")
assertThat(tokenSetSlot.last()).isNull()
```